### ISSUE TITLE
Provide a way to retrieve tags structured by tag type

### ISSUE THREAD

**Issue by CAM-Gerlach** [pre-commit/identify#157]
Currently, `identify` only outputs tags as an unordered set, without any further semantics to allow a consumer to determine which tags correspond to which properties (object type, binary-ness, executable-ness, and file format). For consumption by other tools, it would be very useful to be able to consider each "type" of tag separately; otherwise, to determine the tags in each categories, consumers would need to resort to brittle hardcoding of the which individual tags belong in each category.

Therefore, it would be very useful if there was a means for identify to output its tags structured by type, perhaps as a dictionary with the categories as keys (e.g. `type`, `mode`, `binary`, `format`) as keys and a set of the corresponding tag(s) as the values for each.

To implement this with minimal churn and a backward-compatible public API, the code in `tags_from_path()` could be refactored into another public function, e.g. `get_structured_tags_from_filename()` and add the resulting tags to the set corresponding to the property in question (with an extra line to handle peeling off text and binary into a separate set from the file format parameters, to avoid having to refactor the entirety of `extensions.py`). This function could be used by other packages needing the tags in a structured form, as well as by`tags_from_filename` for concatenation into a single set of tags as current consumers of that function expect.

For some context, I'm in the early stages of building a lightweight open-source tool integrating with pre-commit for managing linter, fixer, checker, and formatter config, and I'm looking to have per-file-type settings (e.g. indentation) inherit properly while using identify's existing standardized tags, i.e. the user sets default settings (indentation, tools to enable, etc) `file`, which can be overriden by `text`, and overriden in turn by e.g. `python` (this particular order being a concern of the consumer, not identify). I could hardcode the tags that belong to each property (and assume the rest are formats), but it would be much more robust and useful if Identify could provide this structured information itself.

I can submit a PR to implement this plus some simple tests, as it should be fairly straightforward to implement, and am also happy to hear further suggestions/recommendations with regard to implementation, how you would suggest structuring the output, etc. Thanks!

**Comment by asottile** [comment]
I'm not sure what you want exactly -- beyond the file types there are only these few tags:

- `symlink` (only symlinks will have this, and no other tags)
- `directory` (in pre-commit's case, only submodules will have this)
- `file` (all files will have this)
- `executable` (it was detected as `X_OK`)
- `non-executable` (it was not detected as `X_OK`)
- `binary` (it is a binary file)
- `text` (it is a text file)

I don't see the benefit in adding structure when there's so few special cases, and 99.9% of the time you really only care about `binary` and `text` which you can test directly for

and if we look at this even more closely, you really only should ever care about `text` and none of the others (you can't "format" a binary file, and formatting symlinks or submodules is nonsense)

so my initial thought is there's no need for this and it would not provide benefit -- perhaps you could demonstrate a use case as to why structure would benefit over what's already there?

**Comment by CAM-Gerlach** [comment]
Hey, sorry for the delay on replying to this.

As a much simpler alternative that wouldn't require changing any of the existing code but avoid the robustness and separation of concerns issues with hardcoding everything consumer-side, we could add a few module-level constants along the lines of the following, simplifying `ALL_TAGS` a bit along the way (names are only tentative), which consumers could use these to construct any structure they like with the returned Identify tags:

```python
OBJECT_TYPES = frozenset({DIRECTORY, FILE, SYMLINK})
MODE_TYPES = frozenset({EXECUTABLE, NON_EXECUTABLE})
ENCODING_TYPES = frozenset({BINARY, TEXT})
FORMAT_TYPES = {
    *extensions.EXTENSIONS.values(),
    *extensions.EXTENSIONS_NEED_BINARY_CHECK.values(),
    *extensions.NAMES.values(),
)
ALL_TAGS = frozenset({*OBJECT_TYPES, *MODE_TYPES, *ENCODING_TYPES, *FORMAT_TYPES })
```

A separate, but much more useful change that can't really be done consumer-side is allowing the _file format_ types to be stored hierarchically, i.e. reflecting that `bashrc` is a sub-type of `bash`, which is a sub-type of `bash`, in turn a sub-type of `shell`, which is a sub-type of `text`. With few exceptions, they are already essentially organized as such in the source of `extensions.py` and `interpreters.py` (and overall, insertion order for the tags is essentially hierarchical). However, the one problem being they are stored as sets, which are of course not order-preserving.

To resolve this, the sets could be converted to lists, which would require major churn (most of it could be automated by regex, but it would still be a pretty comprehensive change) and only converted back to a set on output, with a new function that returned the raw list. A much better alternative would be using [ordered-set](https://github.com/LuminosoInsight/ordered-set) as originally developed by Raymond Hettinger, which is pure Python, has no deps, is on PyPI and conda-forge (with around 60 000 downloads to pre-commit's 260 000), is maintained, and is a drop-in API-compatible replacement for the built-in set that offers the same API (at least everything I tested, covering the operations your code uses), just preserving insertion order.

The only changes required would be a straightforward regex find/replace in `extensions.py` and `interpreters.py`, and changing the four lines in `identify.py` where you instantiate tag sets (40, 42, 44, and 80) to use `OrderedSet`s instead. You _should_ be able to output the `OrderedSet` directly since its API- compatible with regular sets (I've verified it works identically for the places where it is used in pre-commit, as well as Identify itself), but in case you are concerned about potential edge cases, we could cast it back to set on output.

There are a few different usecases I'm considering using `identifile` to handle that would use this:

* The aforementioned one, with cascading settings per type. The example I gave, indentation, isn't a great one without the above change, for the reasons you mention. A more fundamental example would be whether hooks for that type should be enabled or disabled by default, which is relevant for most of the types listed; e.g. you might want to only enable hooks for specific filetypes, so the `enable` value for file would be `false`, but enable checks for symlinks (of which there are two in just the default `pre-commit-hooks` repo) and specific file types, or enable checks for all files (`enable: True` for file), disabling checks for `text` files by default, while enabling them for specific filetypes. A cascading per-language exclude list would be another; e.g. to enable linters, checkers and formatters on general INI files while disabling them on specific INI sub-types.
* This would be used for the hook sort order in the generated pre-commit config file (and perhaps the interactive CLI and elsewhere), so that hooks can be run (and optionally presented, stored, etc) in ascending or descending order of specificity of the the most precise hierarchical level it applies to. This allows e.g. automatically ordering a `pre-commit-config.yaml` so that, e.g., hooks that detect and fix more general problems are run before those aimed at more specific issues, so the e.g. language-specific linters don't need to fix or clutter their output with mundane issues like EOLs, EOFs, tabs/spaces, trailing whitespace, BOM, etc (though this is pretty specific to text files and subtypes thereof), as well as organize the hooks in a more logical order (more generally relevant to all the types).
* I plan to use identify to to suggest or automatically enable and set up the config for just the hooks that apply to the files present. Full-auto setup wouldn't need this (we can either match the individual type/types ourselves or use pre-commit's functions to do that), but for use interactively or simply listing the matched types, it would be useful to be able to list the types hierarchically and determine the most specific type a file has, as well as classify files separately whose most specific type is a given type compared to those that are subtypes (e.g. INI vs Lektorproject or a config file, or text vs. a more specific format).

Hope this helps explain what I'm looking for. Thanks!

**Comment by asottile** [comment]
no thanks

**Comment by CAM-Gerlach** [comment]
I'm sorry; I'm rather confused by your reply. If there are reasons why you feel this is still out of scope for Identify, then I understand, and being one of the maintainers of a large community-developed open-source project (Spyder) myself, it is your decision to not accept a contribution you determine is not in the best interests of the codebase or its users.

However, I'm having trouble grasping why—after you took the time to respond in detail, raised some concerns and requested that I explain further, and I spent several hours carefully considering your detailed points, thinking through and explaining multiple use cases, examining the codebase and proposing an alternative aimed at addressing your points—the only reply would then be a curt two-word rejection, without any reason given. Was there something I unknowingly said or did that was not respectful, appropriate or constructive, despite my best intentions? If so, I apologize.

I would really appreciate if you could let me know what I did wrong, so I can avoid it happening again. I was hoping to make a number of future code contributions to pre-commit and identify, as well as potentially directing a portion of the Spyder organization's financial and developer resources toward supporting its development, considering its positive impact on our own open-source work. However, its hard to move forward with any of that without understanding what went wrong here, to ensure a more productive outcome in the future. Thanks.

**Comment by asottile** [comment]
well for one, because you wrote an essay for what amounts to one maybe two constants was definitely a factor

**Comment by asottile** [comment]
you didn't address any of my points in my original response -- and you're suggesting a full rewrite and a new dependency

**Comment by asottile** [comment]
your requirements suggest new constraints on development which simply don't exist, the tags are unordered and not hierarchical

**Comment by asottile** [comment]
and at the end of the day, you really only need to hard code 2 tags (`text`, `binary`) in your code base to achieve what you want -- suggesting a rewrite and the maintenance burden to shoe-horn a use case that's already trivially solvable on your end isn't going to happen

**Comment by CAM-Gerlach** [comment]
Thanks for explaining; I really appreciate it. I apologize for the verbosity (which I've edited to hopefully reduce), and for my rather misguided conflation of two separate proposals for related but distinct purposes: enumerating the tags by category in constants (straightforward in user code) and ordering all the tags (which is anything but).

For the latter, while I did try to minimize the code changes as much as practicable and enumerate what would specifically be required, its certainly far from trivial, and it seems like its out of scope for Identify. So, it looks like our options for that are either maintain a fork of Identify ourselves, or build a package using a bunch of heuristics and hard-coding to order the Identify output?

For the former, I tried to address your concerns with a much simpler and narrower alternative that didn't impose additional structure on the output or change the code itself, just added a few constants, and described some of the intended use cases beyond text/binary. However, while the constants would simplify Identify's code a bit (a few places already use them) and allow code to consume the output more easily, e.g. `identify_output & identify.FORMAT_TAGS` to get just the filetypes, as you say this part is not vital and not hard to work around if there's a reason not to add them. Is that the case?
